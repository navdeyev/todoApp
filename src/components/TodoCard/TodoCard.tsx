import * as React from 'react';
import styled from 'styled-components';

import {colors} from 'components/Styled/colors';
import {Heading2, Paragraph} from 'components/Styled/Styled';
import {rem} from 'components/Styled/utils';

import {ITodo} from 'domains/types';

const Wrapper = styled.div`
    width: 100%;
    box-sizing: border-box;
    border: 2px solid ${ colors.GREY };
    border-radius: ${ rem(6) };
    margin-bottom: ${ rem(16) };
    padding: ${ rem(20) } ${ rem(16) };
`;
Wrapper.displayName = 'Wrapper';

const HeaderWrapper = styled.div`
    display: flex;
    flex-direction: row;
`;
HeaderWrapper.displayName = 'HeaderWrapper';

const CardHeader = Heading2.extend`flex: 5 1 0px;`;
CardHeader.displayName = 'CardHeader';

const Status = styled.div`
    flex: 1 1 0px;
    display: flex;
    padding: 0 ${ rem(5) }
    align-items: center;
    justify-content: center;
`;
Status.displayName = 'Status';

const Goal = Paragraph.extend`padding-top: ${ rem(8) };`;
Goal.displayName = 'Goal';

export interface IProps {
    todo: ITodo;
}

const TodoCard: React.SFC<IProps> = (props) => {
    const {todo} = props;
    return (
        <Wrapper>
            <HeaderWrapper>
                <CardHeader>{todo.title}</CardHeader>
                <Status>{todo.status}</Status>
            </HeaderWrapper>
            <Goal>{todo.goal}</Goal>
        </Wrapper>
    );
};

export default TodoCard;