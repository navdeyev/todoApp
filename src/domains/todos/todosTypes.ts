export enum TodoStatus {
    PENDING = 'PENDING',
    IN_PROGRESS = 'IN_PROGRESS',
    COMPLETE = 'COMPLETE',
}

export interface ITodoStep {
    id: string,
    title: string,
    details: string
}

export interface ITodo {
    id: string,
    title: string,
    goal: string,
    steps: ITodoStep[],
    status: TodoStatus
}
