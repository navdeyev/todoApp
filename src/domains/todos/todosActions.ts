import {ActionCreator, AnyAction} from 'redux';

import {ITodo} from 'domains/todos/todosTypes';

export enum TodosActions {
    LOAD_TODOS = 'LOAD_TODOS',
    LOAD_TODOS_PENDING = 'LOAD_TODOS_PENDING',
    LOAD_TODOS_SUCCESS = 'LOAD_TODOS_SUCCESS',
    LOAD_TODOS_ERROR = 'LOAD_TODOS_ERROR',

    SELECT_TODO = 'SELECT_TODO',

    UPDATE_STATUS = 'UPDATE_STATUS',
    UPDATE_STATUS_PENDING = 'UPDATE_STATUS_PENDING',
    UPDATE_STATUS_SUCCESS = 'UPDATE_STATUS_SUCCESS',
    UPDATE_STATUS_ERROR = 'UPDATE_STATUS_ERROR',
}

export interface ILoadTodosSuccessAction extends AnyAction {
    payload: ITodo[];
}

const loadTodos: ActionCreator<AnyAction> = () => {
    return {type: TodosActions.LOAD_TODOS};
};

const loadTodosPending: ActionCreator<AnyAction> = () => {
    return {type: TodosActions.LOAD_TODOS_PENDING};
};

const loadTodosSuccess: ActionCreator<ILoadTodosSuccessAction> = (todos: ITodo[]) => {
    return {
        payload: todos,
        type: TodosActions.LOAD_TODOS_SUCCESS
    };
};

const loadTodosError: ActionCreator<AnyAction> = () => {
    return {type: TodosActions.LOAD_TODOS_ERROR};
};

const selectTodo: ActionCreator<AnyAction> = (todoId: string) => {
    return {
        payload: todoId,
        type: TodosActions.SELECT_TODO
    };
};

const updateTodoStatus: ActionCreator<AnyAction> = (todoId: string) => {
    return {
        payload: todoId,
        type: TodosActions.UPDATE_STATUS
    };
};

const updateTodoStatusPending: ActionCreator<AnyAction> = () => {
    return {type: TodosActions.UPDATE_STATUS_PENDING};
};

const updateTodoStatusSuccess: ActionCreator<ILoadTodosSuccessAction> = (todos: ITodo[]) => {
    return {
        payload: todos,
        type: TodosActions.UPDATE_STATUS_SUCCESS
    };
};

const updateTodoStatusError: ActionCreator<AnyAction> = () => {
    return {type: TodosActions.UPDATE_STATUS_ERROR};
};

export default {
    loadTodos,
    loadTodosError,
    loadTodosPending,
    loadTodosSuccess,
    selectTodo,
    updateTodoStatus,
    updateTodoStatusError,
    updateTodoStatusPending,
    updateTodoStatusSuccess
}