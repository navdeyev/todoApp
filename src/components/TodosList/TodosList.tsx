import * as React from 'react';

import {ITodo} from 'domains/todos/todosTypes';

import styled from 'components/Styled/styledComponents';
import {rem} from 'components/Styled/utils';
import TodoCard from 'components/TodoCard/TodoCard';

const Wrapper = styled.div`
    display: flex;
    margin: 0 auto;
    align-items: center;
    box-sizing: border-box;
    padding: 0 ${ rem(20) }
    flex-direction: column;
    max-width: ${ rem(960) };
`;
Wrapper.displayName = 'Wrapper';

export interface IProps {
    todos: ITodo[];
}

const TodosList: React.SFC<IProps> = (props) => {
    const {todos} = props;
    return (
        <Wrapper>
            {todos.map((todo: ITodo) =>
                <TodoCard todo={todo} key={todo.id}/>
            )}
        </Wrapper>
    );
};

export default TodosList;