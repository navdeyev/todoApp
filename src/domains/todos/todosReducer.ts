import * as R from 'ramda';
import {combineReducers} from 'redux';
import {getType} from 'typesafe-actions';

import {LoadingStates} from 'domains/loadingStates';
import {AppAction} from 'domains/rootReducer';
import {ITodo} from 'domains/todos/todosTypes';

import todoActions from './todosActions';

export const todos = (state: ITodo[] = [], action: AppAction) => {
    if (action.type === getType(todoActions.loadTodosSuccess) ||
        action.type === getType(todoActions.updateTodoStatusSuccess)) {
        return action.payload;
    }
    return state;
};

export const loadingState = (state = LoadingStates.NOT_STARTED, action: AppAction) => {
    const pendingArray = [
        getType(todoActions.loadTodosPending),
        getType(todoActions.updateTodoStatusPending),
    ];
    if (R.contains(action.type, pendingArray)) {
        return LoadingStates.LOADING;
    }

    const completeArray = [
        getType(todoActions.loadTodosSuccess),
        getType(todoActions.loadTodosError),
        getType(todoActions.updateTodoStatusSuccess),
        getType(todoActions.updateTodoStatusError)
    ];
    if (R.contains(action.type, completeArray)) {
        return LoadingStates.COMPLETE;
    }

    return state;
};

export const selectedTodoId = (state = '', action: AppAction) => {
    if (action.type === getType(todoActions.selectTodo)) {
        return state === action.payload ? '' : action.payload;
    }
    return state;
};

export default combineReducers({
    loadingState,
    selectedTodoId,
    todos,
});

