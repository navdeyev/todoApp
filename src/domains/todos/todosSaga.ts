import {call, put, takeLatest} from 'redux-saga/effects';

import {IApiService} from "service/api";
import {IServiceMap} from 'service/services';
import {loadTodosError, loadTodosPending, loadTodosSuccess, TodosActions} from './todosActions';

export function* loadTodos(apiService: IApiService) {
    yield put(loadTodosPending());

    try {
        const todos = yield call(apiService.loadTodoList);
        yield put(loadTodosSuccess(todos));
    } catch (e) {
        yield put(loadTodosError());
    }
}

function* todosSaga(serviceMap: IServiceMap) {
    yield takeLatest(TodosActions.LOAD_TODOS, loadTodos, serviceMap.apiService);
}

export default todosSaga;
