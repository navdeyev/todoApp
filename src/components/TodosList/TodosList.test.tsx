import {shallow} from 'enzyme';
import * as React from 'react';

import {default as TodosList, IProps} from './TodosList';

import {TodoStatus} from 'domains/todos/todosTypes';

import TodoCard from 'components/TodoCard/TodoCard';

describe('TogoList', () => {

    let defaultProps: IProps;
    beforeEach(() => {
        defaultProps = {
            todos: [{
                goal: 'First goal description',
                id: 'first-id',
                status: TodoStatus.COMPLETE,
                steps: [],
                title: 'First todo title'
            }, {
                goal: 'Second goal description',
                id: 'second-id',
                status: TodoStatus.COMPLETE,
                steps: [],
                title: 'Second todo title'
            }]
        }
    });

    it('renders a component', () => {
        const render = shallow(<TodosList {...defaultProps} />);
        expect(render.find(TodoCard).length).toBe(2);
        expect(render.getElement()).toMatchSnapshot();
    });

});