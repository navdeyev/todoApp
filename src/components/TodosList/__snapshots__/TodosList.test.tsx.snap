// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TogoList renders a component 1`] = `
<Wrapper>
  <TodoCard
    todo={
      Object {
        "goal": "First goal description",
        "id": "first-id",
        "status": "COMPLETE",
        "steps": Array [],
        "title": "First todo title",
      }
    }
  />
  <TodoCard
    todo={
      Object {
        "goal": "Second goal description",
        "id": "second-id",
        "status": "COMPLETE",
        "steps": Array [],
        "title": "Second todo title",
      }
    }
  />
</Wrapper>
`;
